module/os/linux/zfs/zpl_inode.c

```diff
@@ -149,14 +149,14 @@ zpl_create(struct inode *dir, struct dentry *dentry, umode_t mode, bool flag)
 	error = -zfs_create(ITOZ(dir), dname(dentry), vap, 0,
 	    mode, &zp, cr, 0, NULL);
 	if (error == 0) {
-		d_instantiate(dentry, ZTOI(zp));
-
 		error = zpl_xattr_security_init(ZTOI(zp), dir, &dentry->d_name);
 		if (error == 0)
 			error = zpl_init_acl(ZTOI(zp), dir);
 
 		if (error)
 			(void) zfs_remove(ITOZ(dir), dname(dentry), cr, 0);
+		else
+			d_instantiate(dentry, ZTOI(zp));
```

Comment:
Thanks for catching this, I agree we're going to want to call `d_instantiate` after the ACL init.  However, it looks like we're also going to need to be careful not to leak any of these inodes in the error case.  This is currently happening in this PR which is what caused the test failures.

Right now this is handled by always calling `d_instantiate()` as long as `zfs_create()` (and friends) succeed.  This results in the dentry being added to the dentry hash and then the inode allocated by `zfs_create()` is destroyed in `.destroy_inode->zpl_inode_destroy()->zfs_inode_destroy()`.

So if we're only going to call `d_instantiate()` in the non-failure case here (which I think would be nice).  We'll need to explicitly call `zpl_inode_destroy()` in the error case to avoid leaking that inode.  You can see in the CI logs we can end up hanging in the unmount waiting for these inodes which were leaked to be destroyed.

```
[11249.287059] Call Trace:
[11249.290425]  [<ffffffffc0473f39>] spin_unlock_irqrestore+0x9/0x10 [spl]
[11249.295119]  [<ffffffffc0473f76>] taskq_wait_outstanding_check+0x36/0x40 [spl]
[11249.302621]  [<ffffffffc0473fbf>] taskq_wait_outstanding+0x3f/0xd0 [spl]
[11249.307340]  [<ffffffffc06502d4>] ? dmu_objset_pool+0x24/0x40 [zfs]
[11249.312026]  [<ffffffffc07a4dc1>] zfsvfs_teardown+0x51/0x390 [zfs]
[11249.316683]  [<ffffffffc07a5199>] zfs_umount+0x39/0x130 [zfs]
[11249.321117]  [<ffffffffc07c2934>] zpl_put_super+0x24/0x70 [zfs]
[11249.325583]  [<ffffffff8525069d>] generic_shutdown_super+0x6d/0x100
[11249.330195]  [<ffffffff85250aa2>] kill_anon_super+0x12/0x20
[11249.334588]  [<ffffffffc07c2a7a>] zpl_kill_sb+0x1a/0x20 [zfs]
[11249.339023]  [<ffffffff85250e7e>] deactivate_locked_super+0x4e/0x70
[11249.343658]  [<ffffffff85251606>] deactivate_super+0x46/0x60
[11249.348036]  [<ffffffff85270c8f>] cleanup_mnt+0x3f/0x80
[11249.352288]  [<ffffffff85270d22>] __cleanup_mnt+0x12/0x20
[11249.356589]  [<ffffffff850c28db>] task_work_run+0xbb/0xe0
[11249.360936]  [<ffffffff8502cc65>] do_notify_resume+0xa5/0xc0
```

Reply:
Thank you for the commentary, I wasn't sure exactly how to address the unmounting issue.

1) To ensure I get this right on next force push - is this the correct code fix? (_in pseudo_):
```c
if (error) {
  zfs_remove(...);
  zpl_inode_destroy(ZTOI(zp));
}
else
  d_instantiate(...);
```

2) This should be updated for all 4 functions?
zpl_create()
zpl_mknod()
zpl_mkdir()
zpl_symlink()

Reply:
Yes, that looks like it should work to me.

Reply:
That caused a compile error, should I be using `zfs_inode_destroy()` instead?

Reply:
I agree, it'd be better to keep the `ASSERT`.  You can add `zpl_inode_destroy()` to the zpl_super.c section of the `include/os/linux/zfs/sys/zpl.h` man page to resolve the build failure (right after `zpl_prune_sb()`).  Also `zpl_inode_destroy()` already has a void return type so there's no need to cast it again when called.

Reply:
From the Centos test:
```[ 8838.588490] VERIFY(atomic_read(&ip->i_count) == 0) failed
[ 8838.594872] PANIC at zpl_super.c:47:zpl_inode_destroy()
[ 8838.601101] Showing stack for process 10619
[ 8838.607030] CPU: 0 PID: 10619 Comm: file_write Kdump: loaded Tainted: P           OE  ------------   3.10.0-1160.21.1.el7.x86_64 #1
[ 8838.620139] Hardware name: Amazon EC2 m5d.large/, BIOS 1.0 10/16/2017
[ 8838.627010] Call Trace:
[ 8838.631901]  [<ffffffffa558305a>] dump_stack+0x19/0x1b
[ 8838.638153]  [<ffffffffc05966eb>] spl_dumpstack+0x2b/0x30 [spl]
[ 8838.644759]  [<ffffffffc05967b9>] spl_panic+0xc9/0x110 [spl]
[ 8838.651250]  [<ffffffffc07ef049>] ? spin_unlock+0x9/0x10 [zfs]
[ 8838.657850]  [<ffffffffc07f04eb>] ? rrw_exit+0xcb/0x230 [zfs]
[ 8838.664549]  [<ffffffffc07f08ff>] ? rrm_exit+0x4f/0x90 [zfs]
[ 8838.671131]  [<ffffffffc08cfd01>] ? zfs_remove+0x591/0xb20 [zfs]
[ 8838.677918]  [<ffffffffc08e156a>] zpl_inode_destroy+0x4a/0x50 [zfs]
[ 8838.684809]  [<ffffffffc08e12c3>] zpl_create+0x163/0x200 [zfs]
[ 8838.691375]  [<ffffffffa505ae6a>] vfs_create+0x13a/0x1b0
[ 8838.697671]  [<ffffffffa505eb28>] do_last+0x1168/0x1340
[ 8838.703975]  [<ffffffffc02e4b1c>] ? xfs_filemap_fault+0x2c/0x30 [xfs]
[ 8838.710832]  [<ffffffffa5028b46>] ? kmem_cache_alloc_trace+0x1d6/0x200
[ 8838.717796]  [<ffffffffa505edcd>] path_openat+0xcd/0x5a0
[ 8838.724127]  [<ffffffffa4ff5d50>] ? handle_mm_fault+0xa20/0xfb0
[ 8838.730725]  [<ffffffffa506101d>] do_filp_open+0x4d/0xb0
[ 8838.737141]  [<ffffffffa506f117>] ? __alloc_fd+0x47/0x170
[ 8838.743547]  [<ffffffffa504cb64>] do_sys_open+0x124/0x220
[ 8838.750085]  [<ffffffffa504cc7e>] SyS_open+0x1e/0x20
[ 8838.756302]  [<ffffffffa5595f92>] system_call_fastpath+0x25/0x2a
```
I'm not sure why the Centos 7/8 and Fedora 33 buildbot checks are failing, but the others are succeeding.

Should I be checking for what the ASSERT() will trip on and just not call down to `zpl_inode_destroy()`?  If that is the case, and I'm avoiding the assert trip clause, then should I be calling `zfs_inode_destroy()` directly instead?

Reply:
The ASSERT here indicates we're holding a reference on the inode we're trying to destroy.  Which makes sense since we just set it all up in `zfs_create()->zfs_mknode()->zfs_zoned_alloc()`, what we need to do is unwind that state before destroying it.  That happens when we drop the last reference on the inode via `iput()`.  So looking at this again what I think should work is unhashing the inode then calling `iput()` instead of `zpl_inode_destroy`.  This destroy callback will still run once everything else is cleaned up.

```c
        if (error) {
                zfs_remove(...);
                remove_inode_hash(ZTOI(zp));
                iput(ZTOI(zp));
        } else {
                d_instantiate(...);
        }
```

Looking at this more closely, the ideal place to call `zpl_xattr_security_init()` would really be in `zfs_zoned_alloc()` where we can unwind from a failure more easily.  but the reason this wasn't done initially is because we don't have access to the dentry there, we could pass it in via the `vattr_t` we setup but that's going to be a bigger more disruptive change.  Though it would probably be a good thing to do eventually.

Reply:
Ok - I will implement the `iput()` solution and force push.

Thank you for helping to shepherd this through as we really need SMACK working with ZFS for (podman) container security.

Reply:
4 Failed.  Are they false positive failures?

Reply:
Yes, it look that way.  Those are all known false positives.  I've resubmitted those particular builds to get a clean test run.

Reply:
Is this of concern?
```
Tests with results other than PASS that are unexpected:
    FAIL removal/removal_condense_export (expected PASS)
Error: Process completed with exit code 1.
```
As an aside, any chance of this making it into the current v2.1.0 (rc) series?

Reply:
Not a concern for this PR, it's a false positive we occasionally encounter.

This change is small enough we should be able to get it in to the next rc unless something unexpected crops up after it's been merged to master.